##                        Sneaker Resell Market--SQL analysis
### The purpose of this work is to 1) show some insights of the sneaker resell market; 2) practice some basic and intermediate level sql queries responding some add-hoc analysis needs

### Prework: at the very beginning, shout out to CatherineDevlin's guidance and development! I'd like to use ipython-sql-magic cells to realize running sql and python code at the same time https://github.com/catherinedevlin/ipython-sql

### Some Conclusions:
#### 1) The most market-welcome resell price range is from 200 to 299 dollars, which weights 36.6% of the total sales. Given that the most common release price is 190, I may reach a conclusion that release+(0,100) is the most acceptable ask price range;
#### 2) The market booming happened in 2017 July to Octomber,weighting 45.7% of the three-year-sales;
#### 3) The most popular retro types are Air Jordan 4s,5s,11s,and 13s,counting 40.94% of total sales;
#### 4) The common size for men are from US9.5 to US11,counting 49.69% of transactions;
#### 5) (ad-hoc)  The drving factors of the 2017 Fall Booming can be divided into two parts, one is that 11s SpaceJam, 7210 and 12s The master with good story, good quality and reasonable price; the another one is the 4s and 5s with the average quality but very attractive price (below release price)

In [140]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [141]:
%sql postgresql://postgres:714233@localhost/test
               

u'Connected: postgres@test'

In [3]:
%sql SELECT * FROM pg_user;

#use this query to find the current user of postgresql,
#and the connecting string should follow "postgresql://{user}:{password}@localhost/some_database" 




 * postgresql://postgres:***@localhost/test
1 rows affected.


[(u'postgres', 10, True, True, True, True, u'********', None, None)]

In [175]:
%sql select * from sneakers limit 5;

#this is the completed table after data cleaning and processing
#in the following lines, I will redo this work


 * postgresql://postgres:***@localhost/test
5 rows affected.


[(1, u'Jordan 1 Retro High OG Sail', datetime.date(2017, 10, 17), 10.5, 154, 1),
 (2, u'Jordan 1 Retro High OG Sail', datetime.date(2017, 10, 17), 11.0, 145, 1),
 (3, u'Jordan 1 Retro High OG Sail', datetime.date(2017, 10, 17), 10.5, 200, 1),
 (4, u'Jordan 1 Retro High OG Sail', datetime.date(2017, 10, 17), 8.0, 154, 1),
 (5, u'Jordan 1 Retro High OG Sail', datetime.date(2017, 10, 16), 10.5, 140, 1)]

### 1.1 Create table and Import the csv

In [146]:
%sql drop table sneakers; 
# drop table because it has alreaday been created

 * postgresql://postgres:***@localhost/test
Done.


[]

In [147]:
%sql create table sneakers (id serial primary key,sneaker_name text, \
                            sales_day date,shoe_size varchar(6),price integer,retro_type text);
%sql select * from sneakers; 

 * postgresql://postgres:***@localhost/test
Done.
 * postgresql://postgres:***@localhost/test
0 rows affected.


[]

In [148]:
%sql copy sneakers from 'D:\removal of study material\big data I\final project-sneaker index\data\sneaker complete list.csv' \
delimiter ',' csv header;

 # header here means ignore thr first row of the csv, cuz it is the column name 

 * postgresql://postgres:***@localhost/test
133036 rows affected.


[]

### 2.1 Data Cleaning--Listwise deletion of the null records

In [149]:
%sql select * from sneakers where sneaker_name is null or sales_day is null or shoe_size is null or price is null \
or retro_type is null;

 * postgresql://postgres:***@localhost/test
21 rows affected.


[(6581, u'Jordan 10 Retro Doernbecher', None, u'11', 255, u'10'),
 (9900, u'Jordan 10 Retro Old Royal', None, u'8.5', 146, u'10'),
 (10831, u'Jordan 10 Retro Steel (2005)', None, u'9.5', 175, u'10'),
 (24179, u'Jordan 11 Retro Playoffs CDP (2008)', None, u'8', 460, u'11'),
 (33287, u'Jordan 12 Retro PSNY Wheat', None, u'8', 361, u'12'),
 (33424, u'Jordan 12 Retro Cherry (2009)', None, u'8', 299, u'12'),
 (42411, u'Jordan 13 Retro Low Quai 54', None, u'10', 285, u'13'),
 (43691, u'Jordan 13 Retro Chris Paul Home', None, u'11', 300, u'13'),
 (50625, u'Jordan 14 Retro Light Graphite (2011)', None, None, 96, u'14'),
 (58126, u'Jordan 2 Retro Iron Purple', None, u'9.5', 225, u'2'),
 (67949, u'Jordan 3 Retro Doernbecher (2010)', None, None, 600, u'3'),
 (82129, u'Jordan 4 Retro Military Blue (2006)', None, u'11', 185, u'4'),
 (82563, u'Jordan 4 Retro Fire Red Mars Blackmon', None, u'11', 195, u'4'),
 (96327, u'Jordan 5 Retro Doernbecher', None, u'9.5', 825, u'5'),
 (108798, u'Jordan 6 Retro Brazil World Cup', None, u'10.5', 225, u'6'),
 (108929, u'Jordan 6 Retro Motorsport', None, u'8.5', 400, u'6'),
 (118466, u'Jordan 7 Retro Golden Moments Pack', None, u'11', 270, u'7'),
 (122476, u'Jordan 8 Retro Cool Grey', None, u'8', 186, u'8'),
 (129102, u'Jordan 9 Retro Pinnacle Baseball Black', None, u'13', 260, u'9'),
 (131323, u'Jordan 9 Retro Powder', None, u'7.5', 220, u'9'),
 (132432, u'Jordan 9 Retro White Black Red (2010)', None, u'10.5', 325, u'9')]

In [150]:
%sql delete from sneakers where sneaker_name is null or sales_day is null or shoe_size is null or price is null \
or retro_type is null;

 * postgresql://postgres:***@localhost/test
21 rows affected.


[]

### 2.2 Data Cleaning--Eliminating the PS or GS shoes

In [151]:
%sql select distinct shoe_size from sneakers order by 1 ASC limit 10;
# we find some size following W or C, those are women shoes and baby shoes

 * postgresql://postgres:***@localhost/test
10 rows affected.


[(u'10',),
 (u'10.5',),
 (u'11',),
 (u'11.5',),
 (u'11.5C',),
 (u'11C',),
 (u'11W',),
 (u'12',),
 (u'12.5',),
 (u'12C',)]

In [152]:
%sql select count (*) from sneakers where shoe_size like '%C' or shoe_size like '%W';

 * postgresql://postgres:***@localhost/test
1 rows affected.


[(49L,)]

In [153]:
%sql delete from sneakers where shoe_size like '%C' or shoe_size like '%W';

 * postgresql://postgres:***@localhost/test
49 rows affected.


[]

In [154]:
%sql alter table sneakers alter column shoe_size type float USING shoe_size::double precision;
# after removing the W C sizes, we can modity the column nature to float

 * postgresql://postgres:***@localhost/test
Done.


[]

In [155]:
%sql delete from sneakers where shoe_size < 7;
# remove the GS size, focus on mens size

 * postgresql://postgres:***@localhost/test
801 rows affected.


[]

### 2.3 Data Cleaning--Check retro_type

In [156]:
%sql select distinct retro_type from sneakers; 

 * postgresql://postgres:***@localhost/test
16 rows affected.


[(u'2',),
 (u'4',),
 (u'5',),
 (u'10',),
 (u'7',),
 (u'6',),
 (u'11',),
 (u'14',),
 (u'8',),
 (u'Hy',),
 (u'da',),
 (u'13',),
 (u'9',),
 (u'12',),
 (u'3',),
 (u'1',)]

In [157]:
%sql select *from sneakers where retro_type ='da' or retro_type ='Hy'; 
# based on the result, we need to fix 'da' to 2, and delete 'Hy' row

 * postgresql://postgres:***@localhost/test
18 rows affected.


[(133019, u'Air Jordan 2 Retro QF Black White', datetime.date(2017, 10, 16), 10.5, 140, u'da'),
 (133020, u'Air Jordan 2 Retro QF Black White', datetime.date(2017, 10, 12), 11.0, 200, u'da'),
 (133021, u'Air Jordan 2 Retro QF Black White', datetime.date(2017, 9, 28), 13.0, 150, u'da'),
 (133022, u'Air Jordan 2 Retro QF Black White', datetime.date(2017, 9, 23), 9.0, 300, u'da'),
 (133023, u'Air Jordan 2 Retro QF Black White', datetime.date(2017, 8, 31), 11.5, 275, u'da'),
 (133024, u'Air Jordan 2 Retro QF Black White', datetime.date(2017, 8, 28), 13.0, 250, u'da'),
 (133025, u'Air Jordan 2 Retro QF Black White', datetime.date(2017, 8, 11), 10.0, 250, u'da'),
 (133026, u'Air Jordan 2 Retro QF Black White', datetime.date(2017, 5, 25), 13.0, 250, u'da'),
 (133027, u'Air Jordan 2 Retro QF Black White', datetime.date(2017, 4, 21), 10.0, 300, u'da'),
 (133028, u'Air Jordan 2 Retro QF Black White', datetime.date(2017, 1, 16), 10.5, 265, u'da'),
 (133029, u'Air Jordan 2 Retro QF Black White', datetime.date(2017, 1, 9), 13.0, 250, u'da'),
 (133030, u'Air Jordan 2 Retro QF Black White', datetime.date(2016, 12, 18), 12.0, 300, u'da'),
 (133031, u'Air Jordan 2 Retro QF Black White', datetime.date(2016, 12, 16), 10.0, 250, u'da'),
 (133032, u'Air Jordan 2 Retro QF Black White', datetime.date(2016, 9, 23), 11.5, 275, u'da'),
 (133033, u'Air Jordan 2 Retro QF Black White', datetime.date(2016, 9, 18), 10.0, 265, u'da'),
 (133034, u'Air Jordan 2 Retro QF Black White', datetime.date(2016, 8, 18), 11.0, 300, u'da'),
 (133035, u'Air Jordan 2 Retro QF Black White', datetime.date(2016, 8, 12), 8.5, 300, u'da'),
 (133036, u'Jordan Hyper Violet', datetime.date(2016, 10, 30), 7.5, 175, u'Hy')]

In [158]:
%sql delete from sneakers where retro_type ='Hy';
%sql  update sneakers set retro_type='2' where  retro_type='da';

 * postgresql://postgres:***@localhost/test
1 rows affected.
 * postgresql://postgres:***@localhost/test
17 rows affected.


[]

In [159]:
%sql alter table sneakers alter column retro_type type integer USING retro_type::integer;
# after removing the strange value, we could convert some data type of column

 * postgresql://postgres:***@localhost/test
Done.


[]

### 2.4 Data Cleaning--Working on the prcie range and Detect outliners
#### Based on the diff price range, we get the insight that almost 90% transactions are coming from price $100-499

In [160]:
%sql select sum (case when price between 0 and 99 then 1 else 0 end) as steal, \
 ROUND(100.0 * sum (case when price between 0 and 99 then 1 else 0 end)/COUNT(id),1) as stealper,\
sum (case when price between 100 and 199 then 1 else 0 end) as economic,\
 ROUND(100.0 * sum (case when price between 100 and 199 then 1 else 0 end)/COUNT(id),1) as ecoper, \
sum (case when price between 200 and 299 then 1 else 0 end) as brick,\
 ROUND(100.0 * sum (case when price between 200 and 299 then 1 else 0 end)/COUNT(id),1) as brickper, \
sum (case when price between 300 and 399 then 1 else 0 end) as profit,\
ROUND(100.0 * sum (case when price between 300 and 399 then 1 else 0 end)/COUNT(id),1) as profitper, \
sum (case when price between 400 and 499 then 1 else 0 end) as hype, \
 ROUND(100.0 * sum (case when price between 400 and 499 then 1 else 0 end)/COUNT(id),1) as hypeper, \
sum (case when price between 500 and 599 then 1 else 0 end) as lux, \
 ROUND(100.0 * sum (case when price between 500 and 599 then 1 else 0 end)/COUNT(id),1) as luxper, \
sum (case when price between 600 and 40000 then 1 else 0 end) as super, \
 ROUND(100.0 * sum (case when price between 600 and 40000 then 1 else 0 end)/COUNT(id),1) as superper from sneakers;


 * postgresql://postgres:***@localhost/test
1 rows affected.


[(787L, Decimal('0.6'), 30471L, Decimal('23.1'), 48355L, Decimal('36.6'), 25683L, Decimal('19.4'), 11313L, Decimal('8.6'), 6504L, Decimal('4.9'), 9051L, Decimal('6.8'))]

In [161]:
%sql drop table snkr;
%sql create table snkr as select * from sneakers where price between 100 and 499; 

 * postgresql://postgres:***@localhost/test
Done.
 * postgresql://postgres:***@localhost/test
115822 rows affected.


[]

In [162]:
%sql select * from snkr limit 10;

 * postgresql://postgres:***@localhost/test
10 rows affected.


[(1, u'Jordan 1 Retro High OG Sail', datetime.date(2017, 10, 17), 10.5, 154, 1),
 (2, u'Jordan 1 Retro High OG Sail', datetime.date(2017, 10, 17), 11.0, 145, 1),
 (3, u'Jordan 1 Retro High OG Sail', datetime.date(2017, 10, 17), 10.5, 200, 1),
 (4, u'Jordan 1 Retro High OG Sail', datetime.date(2017, 10, 17), 8.0, 154, 1),
 (5, u'Jordan 1 Retro High OG Sail', datetime.date(2017, 10, 16), 10.5, 140, 1),
 (6, u'Jordan 1 Retro High OG Sail', datetime.date(2017, 10, 16), 8.5, 130, 1),
 (7, u'Jordan 1 Retro High OG Sail', datetime.date(2017, 10, 16), 12.0, 139, 1),
 (8, u'Jordan 1 Retro High OG Sail', datetime.date(2017, 10, 16), 8.0, 135, 1),
 (9, u'Jordan 1 Retro High OG Sail', datetime.date(2017, 10, 16), 11.0, 140, 1),
 (10, u'Jordan 1 Retro High OG Sail', datetime.date(2017, 10, 16), 8.5, 135, 1)]

### 2.5 Data Processing--add column with calculated metrics

In [163]:
#### Adding release price column to the table and calculte the profit
%sql alter table snkr add column release_price integer,add column profit float;

 * postgresql://postgres:***@localhost/test
Done.


[]

In [164]:
%sql update snkr set release_price= 190;
%sql update snkr set release_price= 160 where retro_type=1;
%sql update snkr set release_price= 210 where retro_type=11;
%sql update snkr set release_price= 175 where sneaker_name like'%Low%';
%sql update snkr set release_price= 160 where sneaker_name like'%2%Low%';
%sql update snkr set profit= round(.88*price-release_price,0);

 * postgresql://postgres:***@localhost/test
115822 rows affected.
 * postgresql://postgres:***@localhost/test
1162 rows affected.
 * postgresql://postgres:***@localhost/test
10833 rows affected.
 * postgresql://postgres:***@localhost/test
19248 rows affected.
 * postgresql://postgres:***@localhost/test
2232 rows affected.
 * postgresql://postgres:***@localhost/test
115822 rows affected.


[]

In [165]:
%sql select * from snkr limit 5;

 * postgresql://postgres:***@localhost/test
5 rows affected.


[(4852, u'Jordan 10 Retro Rio', datetime.date(2017, 5, 24), 8.0, 175, 10, 190, -36.0),
 (57683, u'Jordan 2 Retro Melo', datetime.date(2016, 1, 25), 11.0, 350, 2, 190, 118.0),
 (57684, u'Jordan 2 Retro Melo', datetime.date(2016, 1, 13), 8.0, 290, 2, 190, 65.0),
 (60683, u'Jordan 3 Retro Wool', datetime.date(2017, 10, 15), 9.0, 120, 3, 190, -84.0),
 (60684, u'Jordan 3 Retro Wool', datetime.date(2017, 10, 15), 10.0, 120, 3, 190, -84.0)]

#### At this stage, we get the concise dataset named snkr which include records of only men's shoes and exclude the super profitable ones

### 3.1 Data Analytics--Which month has the largest trading volume

In [166]:
%sql with cte as (select distinct to_char(sales_day,'Mon') as month,to_char(sales_day,'YYYY') as year, \
                  count(id) over mont as volume,\
count (id) over () as total, \
round(1.*100 *count(id) over mont/count(id) over(),2) as percent \
from snkr window mont as (partition by date_part('month',sales_day),date_part('year',sales_day)) order by 5 DESC) \
select cte.*,sum(percent) over (order by volume DESC) as cumu_perc from cte limit 15;

 * postgresql://postgres:***@localhost/test
15 rows affected.


[(u'Aug', u'2017', 15513L, 115822L, Decimal('13.39'), Decimal('13.39')),
 (u'Sep', u'2017', 15371L, 115822L, Decimal('13.27'), Decimal('26.66')),
 (u'Oct', u'2017', 11531L, 115822L, Decimal('9.96'), Decimal('36.62')),
 (u'Jul', u'2017', 10517L, 115822L, Decimal('9.08'), Decimal('45.70')),
 (u'Jun', u'2017', 8761L, 115822L, Decimal('7.56'), Decimal('53.26')),
 (u'May', u'2017', 7168L, 115822L, Decimal('6.19'), Decimal('59.45')),
 (u'Dec', u'2016', 5444L, 115822L, Decimal('4.70'), Decimal('64.15')),
 (u'Apr', u'2017', 4944L, 115822L, Decimal('4.27'), Decimal('68.42')),
 (u'Mar', u'2017', 4583L, 115822L, Decimal('3.96'), Decimal('72.38')),
 (u'Feb', u'2017', 4121L, 115822L, Decimal('3.56'), Decimal('75.94')),
 (u'Jan', u'2017', 3418L, 115822L, Decimal('2.95'), Decimal('78.89')),
 (u'Oct', u'2016', 2038L, 115822L, Decimal('1.76'), Decimal('80.65')),
 (u'Aug', u'2016', 1741L, 115822L, Decimal('1.50'), Decimal('82.15')),
 (u'Sep', u'2016', 1713L, 115822L, Decimal('1.48'), Decimal('83.63')),
 (u'Jul', u'2016', 1547L, 115822L, Decimal('1.34'), Decimal('84.97'))]

#### From the result, we know that the most active months are Aug,Sep,Oct,July; in the following section, we will explore the driving factors within those months

### 3.2 Data Analytics--The most popular retro type

In [167]:
%sql with cte as (select distinct retro_type, count(id) over ret, \
                  round(1.*100* count(id) over ret/count(id) over(),2) as percent, \
cast(avg(profit) over ret as decimal(5,0)) as avgprofit \
from snkr window ret as (partition by retro_type)) \
select cte.*, sum(percent) over (order by count DESC) as cumu_perc from cte;

 * postgresql://postgres:***@localhost/test
14 rows affected.


[(5, 14513L, Decimal('12.53'), Decimal('37'), Decimal('12.53')),
 (4, 11917L, Decimal('10.29'), Decimal('54'), Decimal('22.82')),
 (11, 10833L, Decimal('9.35'), Decimal('68'), Decimal('32.17')),
 (13, 10157L, Decimal('8.77'), Decimal('42'), Decimal('40.94')),
 (10, 9610L, Decimal('8.30'), Decimal('14'), Decimal('49.24')),
 (6, 9581L, Decimal('8.27'), Decimal('60'), Decimal('57.51')),
 (12, 9326L, Decimal('8.05'), Decimal('72'), Decimal('65.56')),
 (7, 8929L, Decimal('7.71'), Decimal('29'), Decimal('73.27')),
 (3, 8437L, Decimal('7.28'), Decimal('37'), Decimal('80.55')),
 (9, 7210L, Decimal('6.23'), Decimal('29'), Decimal('86.78')),
 (8, 5029L, Decimal('4.34'), Decimal('11'), Decimal('91.12')),
 (2, 4767L, Decimal('4.12'), Decimal('-12'), Decimal('95.24')),
 (14, 4351L, Decimal('3.76'), Decimal('32'), Decimal('99.00')),
 (1, 1162L, Decimal('1.00'), Decimal('-6'), Decimal('100.00'))]

#### Now we know the market perfers retro 5, 4, 11, 13 over others

### 3.3 Data Analytics--Which size is the most popular

In [168]:
%sql with cte as (select distinct shoe_size, count(id) over soe, \
                  round(1.*100* count(id) over soe/count(id) over(),2) as percent, \
cast(avg(profit) over soe as decimal(5,0)) as avgprofit \
from snkr window soe as (partition by shoe_size)) \
select cte.*,sum(percent) over (order by count DESC) as cumu_percent from cte;

 * postgresql://postgres:***@localhost/test
20 rows affected.


[(11.0, 15536L, Decimal('13.41'), Decimal('41'), Decimal('13.41')),
 (10.5, 15346L, Decimal('13.25'), Decimal('39'), Decimal('26.66')),
 (10.0, 14979L, Decimal('12.93'), Decimal('38'), Decimal('39.59')),
 (9.5, 11700L, Decimal('10.10'), Decimal('43'), Decimal('49.69')),
 (12.0, 11497L, Decimal('9.93'), Decimal('42'), Decimal('59.62')),
 (9.0, 9718L, Decimal('8.39'), Decimal('39'), Decimal('68.01')),
 (13.0, 7655L, Decimal('6.61'), Decimal('44'), Decimal('74.62')),
 (11.5, 7183L, Decimal('6.20'), Decimal('42'), Decimal('80.82')),
 (8.5, 6858L, Decimal('5.92'), Decimal('39'), Decimal('86.74')),
 (8.0, 5492L, Decimal('4.74'), Decimal('36'), Decimal('91.48')),
 (14.0, 3876L, Decimal('3.35'), Decimal('44'), Decimal('94.83')),
 (7.5, 1577L, Decimal('1.36'), Decimal('30'), Decimal('96.19')),
 (7.0, 1219L, Decimal('1.05'), Decimal('8'), Decimal('97.24')),
 (12.5, 1206L, Decimal('1.04'), Decimal('39'), Decimal('98.28')),
 (15.0, 1034L, Decimal('0.89'), Decimal('29'), Decimal('99.17')),
 (13.5, 332L, Decimal('0.29'), Decimal('45'), Decimal('99.46')),
 (16.0, 301L, Decimal('0.26'), Decimal('45'), Decimal('99.72')),
 (17.0, 180L, Decimal('0.16'), Decimal('57'), Decimal('99.88')),
 (18.0, 131L, Decimal('0.11'), Decimal('44'), Decimal('99.99')),
 (14.5, 2L, Decimal('0.00'), Decimal('-36'), Decimal('99.99'))]

#### This result shows a range between 9.5 and 11, confirms the academic finding that nowadays the average shoe size of american men is US10.

### 3.4 Data Analytics--(ad-hoc)Which shoes most contribute to the high trading volume during hype months

In [169]:
%sql with cte as (select distinct retro_type,count(id) over (partition by retro_type) as count,\
round(1.*100*count(id) over (partition by retro_type)/count(id) over (),2) as percent from snkr \
where date_part('month',sales_day) in (7,10) and date_part('year',sales_day) =2017) \
select cte.*, sum(percent) over (order by count DESC) as cumu_percent from cte;

 * postgresql://postgres:***@localhost/test
14 rows affected.


[(11, 2710L, Decimal('12.29'), Decimal('12.29')),
 (12, 2607L, Decimal('11.82'), Decimal('24.11')),
 (5, 2558L, Decimal('11.60'), Decimal('35.71')),
 (4, 2556L, Decimal('11.59'), Decimal('47.30')),
 (6, 1969L, Decimal('8.93'), Decimal('56.23')),
 (13, 1921L, Decimal('8.71'), Decimal('64.94')),
 (7, 1670L, Decimal('7.57'), Decimal('72.51')),
 (3, 1486L, Decimal('6.74'), Decimal('79.25')),
 (10, 1269L, Decimal('5.76'), Decimal('85.01')),
 (8, 929L, Decimal('4.21'), Decimal('89.22')),
 (9, 876L, Decimal('3.97'), Decimal('93.19')),
 (14, 545L, Decimal('2.47'), Decimal('95.66')),
 (1, 543L, Decimal('2.46'), Decimal('98.12')),
 (2, 409L, Decimal('1.86'), Decimal('99.98'))]

#### From the result, we may conclude that 11,12,4,5 contributed the 47.3% of sales within July to October period; in the following section, we will explore the certain ones

In [170]:
%sql select distinct sneaker_name, count(id) over (partition by sneaker_name),\
count(id) over() as total, round(1.*100*count(id) over (partition by sneaker_name)/count(id) over(),2) as percent, \
cast(avg(profit) over(partition by sneaker_name) as integer) as avgprofit, \
cast(avg(price) over(partition by sneaker_name) as integer) as avgprice \
from snkr where date_part('month',sales_day) in (7,10) and date_part('year',sales_day) =2017 \
and retro_type=11 order by 2 DESC limit 5;

 * postgresql://postgres:***@localhost/test
5 rows affected.


[(u'Jordan 11 Retro Space Jam (2016)', 240L, 2710L, Decimal('8.86'), 28, 270),
 (u'Jordan 11 Retro 72-10', 226L, 2710L, Decimal('8.34'), 46, 291),
 (u'Jordan 11 Retro Low Cherry (2016)', 220L, 2710L, Decimal('8.12'), 34, 238),
 (u'Jordan 11 Retro Low University Blue (2017)', 206L, 2710L, Decimal('7.60'), 3, 202),
 (u'Jordan 11 Retro Low Bred', 197L, 2710L, Decimal('7.27'), 109, 323)]

In [171]:
%sql select distinct sneaker_name, count(id) over (partition by sneaker_name),\
count(id) over() as total, round(1.*100*count(id) over (partition by sneaker_name)/count(id) over(),2) as percent, \
cast(avg(profit) over(partition by sneaker_name) as integer) as avgprofit, \
cast(avg(price) over(partition by sneaker_name) as integer) as avgprice \
from snkr where date_part('month',sales_day) in (7,10) and date_part('year',sales_day) =2017 \
and retro_type=12 order by 2 DESC limit 5;

 * postgresql://postgres:***@localhost/test
5 rows affected.


[(u'Jordan 12 Retro The Master', 236L, 2607L, Decimal('9.05'), 100, 329),
 (u'Jordan 12 Retro Deep Royal Blue', 221L, 2607L, Decimal('8.48'), 44, 265),
 (u'Jordan 12 Retro Gym Red', 221L, 2607L, Decimal('8.48'), 55, 278),
 (u'Jordan 12 Retro Bordeaux', 213L, 2607L, Decimal('8.17'), -24, 188),
 (u'Jordan 12 Retro Wool', 205L, 2607L, Decimal('7.86'), 25, 244)]

In [172]:
%sql select distinct sneaker_name, count(id) over (partition by sneaker_name),\
count(id) over() as total, round(1.*100*count(id) over (partition by sneaker_name)/count(id) over(),2) as percent, \
cast(avg(profit) over(partition by sneaker_name) as integer) as avgprofit, \
cast(avg(price) over(partition by sneaker_name) as integer) as avgprice \
from snkr where date_part('month',sales_day) in (7,10) and date_part('year',sales_day) =2017 \
and retro_type=4 order by 2 DESC limit 5;

 * postgresql://postgres:***@localhost/test
5 rows affected.


[(u'Jordan 4 Retro Pure Money (2017)', 313L, 2556L, Decimal('12.25'), -26, 186),
 (u'Jordan 4 Retro Oreo (2015)', 214L, 2556L, Decimal('8.37'), 22, 241),
 (u'Jordan 4 Retro Alternate 89', 202L, 2556L, Decimal('7.90'), -14, 200),
 (u'Jordan 4 Retro Royalty', 169L, 2556L, Decimal('6.61'), 9, 226),
 (u'Jordan 4 Retro White Cement (2016)', 161L, 2556L, Decimal('6.30'), 57, 281)]

In [173]:
%sql select distinct sneaker_name, count(id) over (partition by sneaker_name),\
count(id) over() as total, round(1.*100*count(id) over (partition by sneaker_name)/count(id) over(),2) as percent, \
cast(avg(profit) over(partition by sneaker_name) as integer) as avgprofit, \
cast(avg(price) over(partition by sneaker_name) as integer) as avgprice \
from snkr where date_part('month',sales_day) in (7,10) and date_part('year',sales_day) =2017 \
and retro_type=5 order by 2 DESC limit 5;

 * postgresql://postgres:***@localhost/test
5 rows affected.


[(u'Jordan 5 Retro Blue Suede', 345L, 2558L, Decimal('13.49'), -16, 198),
 (u'Jordan 5 Retro Bronze', 303L, 2558L, Decimal('11.85'), -16, 198),
 (u'Jordan 5 Retro Olympic (2016)', 197L, 2558L, Decimal('7.70'), -20, 194),
 (u'Jordan 5 Retro Red Suede', 136L, 2558L, Decimal('5.32'), -10, 204),
 (u'Jordan 5 Retro Black Metallic (2016)', 134L, 2558L, Decimal('5.24'), 47, 270)]

In [174]:
%sql select distinct sneaker_name,count(sneaker_name) from snkr where date_part('year',sales_day)=2017 \
and date_part('month',sales_day) in (7,10) and retro_type in (4,5,11,12) group by sneaker_name order by 2 DESC limit 10;

 * postgresql://postgres:***@localhost/test
10 rows affected.


[(u'Jordan 5 Retro Blue Suede', 345L),
 (u'Jordan 4 Retro Pure Money (2017)', 313L),
 (u'Jordan 5 Retro Bronze', 303L),
 (u'Jordan 11 Retro Space Jam (2016)', 240L),
 (u'Jordan 12 Retro The Master', 236L),
 (u'Jordan 11 Retro 72-10', 226L),
 (u'Jordan 12 Retro Deep Royal Blue', 221L),
 (u'Jordan 12 Retro Gym Red', 221L),
 (u'Jordan 11 Retro Low Cherry (2016)', 220L),
 (u'Jordan 4 Retro Oreo (2015)', 214L)]

#### Easily to see, within the 11s group 7210, SpaceJam, University Blue, Cherry and Bred are all selling at premium, in the 12s group, things are much same. However, the majority of 4s and 5s groups are on-sales or clearance ones.
#### Based on my knowledge, the 11s and 12s belong to the hype level, of which resell price is profitable, however, the 4s and 5s are economy choices, of which resell is below the release price or lean.